CI update

j-stephan opened the pull request:
Summary:

* ~~added gcc-12 to CI~~ - handled in #1765
* ~~added clang-14 to CI~~ - handled in #1766
* ~~added Xcode 13.3.1 to CI -> we now support macOS 12 "Monterey"~~ - handled in #1759 
* ~~added CUDA 11.7 to CI~~ - handled in #1890
* ~~CUDA 11.6 tests now use CUDA 11.6.2~~ - handled in #1890
* ~~collapsed compiler support matrix~~ - handled in #1860 
* ~~fixed a few consistency issues for the clang-cuda device compiler (the matrix showed support for CUDA versions that clang itself doesn't officially support yet)~~ - handled in #1890
* ~~added CUDA 11.6 + gcc-11 job to GitLab CI~~ - handled in #1890
* ~~added ROCm 5.1 jobs to GitLab CI~~ - handled in #1762.
* ~~added Boost 1.79.0 to CI~~ - handled in #1903
* ~~added CMake 3.23.x to CI~~ - handled in #1903
* ~~updated older CMake versions to latest point releases~~ - handled in #1903
* ~~removed all occurrences of `alpaka_CXX_STANDARD: 17` from the GitHub CI script since that is now our default~~ - handled in #1902
* ~~added a few occurrences of `alpaka_CXX_STANDARD: 20` to the GitHub CI script~~ - handled in #1902
* ~~gcc-12 jobs will use GitHub's `ubuntu-22.04` runners~~ - handled in #1765
* ~~If sanitizers are enabled, clang jobs will install the llvm base package. This includes the `llvm-symbolizer` binary required for meaningful output.~~ - handled in #1763 
* ~~If the Boost.fiber back-end is enabled in C++20 mode CMake will print an error and abort (see #1714).~~ - superseded by #1718
* ~~Disabled the address sanitizer for `memBufTest` if clang-14 is used (see #1753).~~ - needs more investigation
* ~~Removed Xcode 12.4.0 from the CI -> we will no longer support macOS 10.15 "Catalina"~~ - handled in #1759.

Fixes #1670. Fixes #1708. Fixes #1724.

j-stephan committed:
CI update

j-stephan commented:
Wow, case in point: One of our C++20 jobs is failing with Boost 1.79.0. Apparently Boost.fiber is internally using `std::result_of` which was removed in C++20.

From Boost.fiber's documentation:

> Boost.fiber requires C++11!

So I guess we need to start thinking about the future of this back-end.

bernhardmgruber commented:
> Wow, case in point: One of our C++20 jobs is failing with Boost 1.79.0. Apparently Boost.fiber is internally using `std::result_of` which was removed in C++20.

That was fixed 15 days ago: https://github.com/boostorg/fiber/commit/a6b64c9a69d08d030cf1533319e670f70ce6585e

However, that still leaves Boost.Fiber 1.79 and below problematic with C++20.

bernhardmgruber commented:
It seems `g++-12` is not installed:
```bash
2022-05-10T17:06:01.9941211Z +++ sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-12 50
2022-05-10T17:06:01.9987407Z update-alternatives: error: alternative path /usr/bin/g++-12 doesn't exist
```

j-stephan commented:
You're right. `apt install g++-12` leads to an installation of clang-12 for some reason. Maybe it cannot find g++-12 and does some guesswork. :clown_face:

j-stephan commented:
Rebased and added the following changes:

* added CUDA 11.7 to CI
* CUDA 11.6 tests now use CUDA 11.6.2
* gcc-12 jobs will use GitHub's ubuntu-22.04 runners

j-stephan commented:
* gcc-12 jobs are failing because there is a Docker bug in the ubuntu-22.04 runners. This has been fixed but is apparently still in the rollout phase.
* clang-14-asan is failing because of a bug in one of our test cases. We need to investigate this.

jkelling commented:
While you are at it: Could you please change the build type of check with enabled sanitizer which are currently `Release` to `RelWithDebInfo` to make the printed backtraces more useful? (e.g. `linux_clang-6_release_asan`, I do not know if there are others)

j-stephan commented:
> While you are at it: Could you please change the build type of check with enabled sanitizer which are currently Release to RelWithDebInfo to make the printed backtraces more useful?

Done.

j-stephan commented:
> clang-14-asan is failing because of a bug in one of our test cases. We need to investigate this.

It looks like this is caused by Catch2 instead of our code. Hopefully this will be fixed once we migrate to Catch2 v3.

j-stephan commented:
Note: The macOS 10.15 runners have been deprecated by GitHub and will be removed at the end of August. I therefore removed our macOS 10.15 tests from the CI which will be visible upon my next push (after merging #1754 and #1755).

psychocoderHPC commented:
I am a little bit unhappy that this PR is mixing features, refactorings, and adding new test combinations.

Could you split e.g CUDA keyring refactorings, job renamings and new jobs into separate PRs? This would also simplify the review and give us a better overview of why a CI job is failing.

j-stephan commented:
And so it shall be done. I will prepare a number of successor PRs and close this one afterwards.

bernhardmgruber commented on script/install_clang.sh:
```diff
@@ -36,8 +36,35 @@ then
     sudo add-apt-repository 'deb http://apt.llvm.org/focal/ llvm-toolchain-focal-13 main'
 fi
 
+# add clang-13 repository for ubuntu 22.04
+if [[ "$(cat /etc/os-release)" == *"22.04"* && "${ALPAKA_CI_CLANG_VER}" -eq 13 ]]
+then
+    wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
+    sudo add-apt-repository 'deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-13 main'
+fi
+
+# add clang-14 repository for ubuntu 20.04
+if [[ "$(cat /etc/os-release)" == *"20.04"* && "${ALPAKA_CI_CLANG_VER}" -eq 14 ]]
+then
+    wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
+    sudo add-apt-repository 'deb http://apt.llvm.org/focal/ llvm-toolchain-focal-14 main'
+fi
+
+# add clang-14 repository for ubuntu 22.04
+if [[ "$(cat /etc/os-release)" == *"22.04"* && "${ALPAKA_CI_CLANG_VER}" -eq 14 ]]
+then
+    wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
+    sudo add-apt-repository 'deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-14 main'
+fi
+
```
I am by no means skilled in bash. Could we have something like that:
```suggestion
codename=`lsb_release -c  | awk '{print $2}'`
wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
sudo add-apt-repository "deb http://apt.llvm.org/$codename/ llvm-toolchain-$codename-$ALPAKA_CI_CLANG_VER main"
```

j-stephan replied:
Thanks for the suggestion, I picked this up in #1763.

bernhardmgruber commented on script/install_clang.sh:
```diff
@@ -36,8 +36,35 @@ then
+
 travis_retry sudo apt-get -y --quiet --allow-unauthenticated --no-install-recommends install clang-${ALPAKA_CI_CLANG_VER}
 
+if [ -n "${ALPAKA_CI_SANITIZERS}" ]
+then
+    # llvm-symbolize is required for meaningful output. This is part of the llvm base package which we don't install by default.
+    travis_retry sudo apt-get -y --quiet --allow-unauthenticated --no-install-recommends install llvm-${ALPAKA_CI_CLANG_VER}
+fi
+
```
I hope installing LLVM does not add a lot of time to the CI setup time. Isn't there a sub-package we could get?

j-stephan replied:
Unfortunately the `llvm-symbolizer` is only part of the base package. :-(

bernhardmgruber replied:
Ok, let's see how that affects us. We only need it for the sanitizer builds though, so it should only affect a small number of jobs.

j-stephan commented:
All points from the original description have been addressed.
